torchtune/models/qwen2/_prompt_template.py

```diff
@@ -0,0 +1,117 @@
+
+
+class Qwen25ChatTemplate(PromptTemplateInterface):
+    """Qwen2.5's chat template.
+
+    Defined in the Jinja template in
+    https://huggingface.co/Qwen/Qwen2.5-7B-Instruct/raw/main/tokenizer_config.json
+
+    Attributes:
+        template: a dictionary mapping role to the
+            prepend and append tags
+
+    Args:
+        add_default_sys_prompt_if_missing (bool): Add Qwen's default system prompt if the messages do not have one.
+            Default: False
+        tools (Optional[List[str]]): Optional JSON schemas for tools that the LLM can access.
+            If provided, the tool schemas will be added to the system prompt according to Qwen's tool instruction format.
+            See https://huggingface.co/blog/unified-tool-use for more information about the JSON tool schema.
+            Default: None
+    """
+
+    template = {
+        "system": ("<|im_start|>system\n", "<|im_end|>\n"),
+        "user": ("<|im_start|>user\n", "<|im_end|>\n"),
+        "assistant": ("<|im_start|>assistant\n", "<|im_end|>\n"),
+    }
+
+    def __init__(
+        self,
+        add_default_sys_prompt_if_missing: bool = False,
+        tools: Optional[List[str]] = None,
+    ):
+        self._add_default_sys_prompt_if_missing = add_default_sys_prompt_if_missing
+        self._tools = tools
+
+    def __call__(
+        self,
+        messages: List[Message],
+        inference: bool = False,
+    ) -> List[Message]:
+        """
+        Format user, assistant, and system messages with appropriate tags.
+
+        Args:
+            messages (List[Message]): a single conversation, structured as a list
+                of `Message` objects
+            inference (bool): Whether the template is being used for inference or not.
+
+        Returns:
+            The formatted list of messages
+        """
+        # Add a system prompt if missing
+        if self._add_default_sys_prompt_if_missing and (
+            not messages or messages[0].role != "system"
+        ):
+            messages.insert(
+                0, Message(role="system", content=DEFAULT_SYS_PROMPT, masked=True)
+            )
+
+        # Add tool instructions to the system prompt
+        if self._tools and messages[0].role == "system":
+            tool_instruction = [TOOL_INSTRUCTION_START]
+            for tool in self._tools:
+                tool_instruction.append(tool)
+            tool_instruction.append(TOOL_INSTRUCTION_END)
+            tool_instruction = "\n".join(tool_instruction)
+            assert messages[0].role == "system"
+            messages[0].content.append({"type": "text", "content": tool_instruction})
+
+        # Add start/end tags to messages
+        formatted_dialogue = []
+        for i, message in enumerate(messages):
```

Comment:
is the primary logic here for adding tool call tokens? but these are special tokens correct? so adding them as strings may not get tokenized correctly. Instead, special tokens should be handled by `tokenize_messages` in the tokenizer. We should reserve the prompt template for non-special token text that is used for Qwen instruct, maybe the prompts you included at the top.

Before, it was just the ChatMLTemplate using `<|im_start|>` etc. If these have become special tokens then we should move that logic to the tokenizer. Otherwise, we should leave the logic the same as before, except maybe including the default system prompt you added.

I am not sure yet how to handle registered/available tools...

Reply:
Adding them as strings does result in them being tokenized. `tokenizer.tokenize_messages` calls `tokenizer.encode` which detects these strings (any keys of the special tokens dict) and splits them out and converts them to the token IDs.

I think leaving this logic in the prompt template is better than moving it into `tokenize_messages` because we enable the user to pass in their own custom template. If we hide that logic in the tokenizer, they won't be able to customize it, and it would make the tokenizer more fragile (it could break if the user uses a custom template and the tokenizer receives messages in an unexpected format).

Reply:
I see your point about configurability, in this case I am considering "prompt template" just to be any task related normal text, like "Question: Q Answer: A" that you want to structure inputs into. This is handled in the prompt template classes. Anything that has their own reserved token id should be handled in the tokenizer because they need to be added for that model regardless of what the task template is. This is to minimize accidental user error.

My main concern with this is that we're not being consistent here. And for configurability, we should probably agree on where users should configure custom special tokens. Right now model tokenizers have the option to modify that in the tokenizers builder (or maybe we need to make the experience better). There is also the prompt template which is a separate place to configure the actual task template (maybe we should name it that to avoid confusion). This changes kinda conflates the two and it's not clear to me where I should go for either case.

Reply:
Just getting caught up on this.. I think both of you have good points here. Ideally we would just be able to support the use case @calvinpelletier is describing just by changing the prompt template. But I definitely would prefer to address that in a systematic way rather than doing just a one-off here. Iiuc `<tool_response>` and `<\tool_response>` are both recognized by the tokenizer, so given that I don't think it makes sense to use them in string form here (plus imo doing this is also kinda brittle). Re the point about not being able to customize if we put this in the tokenizer, I'm not sure I 100% follow. What kind of customization would not be supported? Even in the worst case, one can inherit the tokenizer class and just override `tokenize_messages` as needed (which again, if users accept our claim of "special tokens get dealt with in the tokenizer" may be a bit more friction, but is at least the expected thing to do).

Reply:
Since the prompt template already adds special tokens to the messages (`<|im_start|>`, etc.) , my understanding was that the prompt template is responsible for adding special token _text_ and the tokenizer is responsible for converting the text into token IDs. Wrapping an ipython message in `<tool_response>` feels like the same logic as wrapping a user message in `<|im_start|>user`.

That's a fair point about the customization though. At first, I was thinking a user might want to do things like have multiple tool responses in the same message or add additional XML tags like `<tool_status>` or whatever. But now that I'm thinking more about this, I don't think it's something we need to worry about.

Reply:
ironically, the <|im_start|> is not a reserved token ID. It's just raw text that gets tokenized normally that can be used by any model, hence why it's located in prompt template and not in the tokenizer.

> my understanding was that the prompt template is responsible for adding special token text and the tokenizer is responsible for converting the text into token IDs.

yeah maybe we should make this more clear. in my mind the distinction was prompt template = structured plain text, tokenizer = all special tokens. we can discuss if we want to change this but I would say to be consistent with this until we do make any changes.